internal/transform/contract_data.go

```diff
@@ -0,0 +1,503 @@
+func (t *TransformContractDataStruct) TransformContractData(ledgerChange ingest.Change, passphrase string) (ContractDataOutput, error) {
+	ledgerEntry, changeType, outputDeleted, err := utils.ExtractEntryFromChange(ledgerChange)
+	if err != nil {
+		return ContractDataOutput{}, err
+	}
+
+	contractData, ok := ledgerEntry.Data.GetContractData()
+	if !ok {
+		return ContractDataOutput{}, fmt.Errorf("Could not extract contract data from ledger entry; actual type is %s", ledgerEntry.Data.Type)
+	}
+
+	// LedgerEntryChange must contain a contract data change to be parsed, otherwise skip
+	if ledgerEntry.Data.Type != xdr.LedgerEntryTypeContractData {
+		return ContractDataOutput{}, nil
+	}
+
+	contractDataAssetCode, contractDataAssetIssuer := t.AssetFromContractData(ledgerEntry, passphrase)
+
+	contractDataBalanceHolder, contractDataBalance, _ := t.ContractBalanceFromContractData(ledgerEntry, passphrase)
+
+	if contractData.Key.Type.String() == "ScValTypeScvLedgerKeyNonce" {
+		// Is a nonce and should be discarded
+		return ContractDataOutput{}, nil
+	}
+
+	contractDataContractId, ok := contractData.Contract.GetContractId()
+	if !ok {
+		return ContractDataOutput{}, fmt.Errorf("Could not extract contractId data information from contractData")
+	}
+
+	contractDataKeyType := contractData.Key.Type.String()
+
+	//keyBinary, err := contractData.Key.MarshalBinary()
+	//if err != nil {
+	//	return ContractDataOutput{}, fmt.Errorf("Could not extract Key from contractData")
+	//}
+	//contractDataKey := base64.StdEncoding.EncodeToString(keyBinary)
+
+	contractDataDurability := contractData.Durability.String()
+
+	contractDataData, ok := contractData.Body.GetData()
+	if !ok {
+		return ContractDataOutput{}, fmt.Errorf("Could not extract contract data information from contractId %s", xdr.Hash(*contractData.Contract.ContractId).HexString())
+	}
+	contractDataDataFlags := contractDataData.Flags
+
+	// Useful contract data values are flattened into their own columns
+	//valBinary, err := contractDataData.Val.MarshalBinary()
+	//if err != nil {
+	//	return ContractDataOutput{}, fmt.Errorf("Could not extract Val from contractData")
+	//}
+	//contractDataDataVal := base64.StdEncoding.EncodeToString(valBinary)
+
+	contractDataExpirationLedgerSeq := contractData.ExpirationLedgerSeq
+
+	transformedPool := ContractDataOutput{
+		ContractId:                  contractDataContractId.HexString(),
+		ContractKeyType:             contractDataKeyType,
+		ContractDurability:          contractDataDurability,
+		ContractDataFlags:           uint32(contractDataDataFlags),
+		ContractExpirationLedgerSeq: uint32(contractDataExpirationLedgerSeq),
+		ContractDataAssetCode:       contractDataAssetCode,
+		ContractDataAssetIssuer:     contractDataAssetIssuer,
+		ContractDataBalanceHolder:   base64.StdEncoding.EncodeToString(contractDataBalanceHolder[:]),
+		ContractDataBalance:         contractDataBalance.String(),
```

Comment:
do we want to store the balance as a string?

Reply:
Balance needs to be a string I think because it is the `big.Int` type in golang. We would need to convert it to `BIGNUMERIC` in BQ when we ingest it